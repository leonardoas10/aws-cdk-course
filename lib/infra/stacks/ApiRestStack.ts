import { Stack, StackProps } from 'aws-cdk-lib';
import { LambdaIntegration, RestApi } from 'aws-cdk-lib/aws-apigateway';
import { Construct } from 'constructs';

type THttpMethod =
    | 'GET'
    | 'POST'
    | 'PUT'
    | 'DELETE'
    | 'PATCH'
    | 'OPTIONS'
    | 'HEAD';

interface IResourceMethodConfig {
    path: string;
    methods: THttpMethod[];
    lambdaIntegration: LambdaIntegration;
}

interface IApiRestStackProps extends StackProps {
    readonly apiGatewayName: string;
    readonly resourceConfigs: IResourceMethodConfig[];
    readonly apiGatewayDescription?: string;
}

export class ApiRestStack extends Stack {
    constructor(
        scope: Construct,
        id: string,
        {
            apiGatewayName,
            apiGatewayDescription = '<insert a properly description>',
            resourceConfigs,
            ...props
        }: IApiRestStackProps
    ) {
        super(scope, id, props);

        const api = new RestApi(this, apiGatewayName, {
            restApiName: apiGatewayName,
            description: apiGatewayDescription,
        });

        resourceConfigs.forEach((config) => {
            const resource = api.root.addResource(config.path);
            config.methods.forEach((method) => {
                resource.addMethod(method, config.lambdaIntegration);
            });
        });
    }
}
