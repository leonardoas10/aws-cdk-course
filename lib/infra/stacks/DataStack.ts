import { Stack, StackProps } from 'aws-cdk-lib';
import {
    Table,
    TableProps,
    AttributeType,
    ITable,
} from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';
import { getSuffixFromStack } from '../shared/Utils';

/**
 * The `DataStack` class is a custom AWS CDK stack that provisions a DynamoDB table.
 *
 * This stack creates a DynamoDB table with a partition key named 'id' of type `STRING`.
 * The table's name is derived from the stack ID combined with a suffix generated by the
 * `getSuffixFromStack` utility function.
 *
 * @param scope The scope in which this stack is defined.
 * @param id The identifier for this stack.
 * @param props The properties for configuring the stack, including:
 *  - Any additional properties from `StackProps` that might be needed for configuring the stack.
 */
export class DataStack extends Stack {
    public readonly table: ITable;

    constructor(scope: Construct, id: string, { ...props }: StackProps) {
        super(scope, id, props);

        const suffix = getSuffixFromStack(this);

        this.table = new Table(this, id, {
            partitionKey: { name: 'id', type: AttributeType.STRING },
            tableName: `${id}-${suffix}`,
        } as TableProps);
    }
}
