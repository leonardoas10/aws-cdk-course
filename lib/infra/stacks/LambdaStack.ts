import { Stack, StackProps, Duration } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { join } from 'path';
import { LambdaIntegration } from 'aws-cdk-lib/aws-apigateway';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Effect, PolicyStatement } from 'aws-cdk-lib/aws-iam';

interface LambdaStackProps extends StackProps {
    readonly lambdaName: string;
    readonly entryPoint: string;
    readonly policyActions?: Array<string>;
    readonly policiyResources?: Array<string>;
    readonly environment?: { [key: string]: string };
    readonly runtime?: Runtime;
    readonly timeout?: number;
}

export class LambdaStack extends Stack {
    public readonly lambdaIntegration: LambdaIntegration;
    public readonly lambda: NodejsFunction;

    constructor(
        scope: Construct,
        id: string,
        {
            lambdaName,
            entryPoint = '/',
            environment = {},
            runtime = Runtime.NODEJS_20_X,
            policyActions = ['*'],
            policiyResources = ['*'],
            timeout = 30,
            ...props
        }: LambdaStackProps
    ) {
        super(scope, id, props);

        const lambda = new NodejsFunction(this, lambdaName, {
            runtime: runtime,
            entry: join(__dirname, `../../${entryPoint}`),
            handler: 'handler',
            environment,
            timeout: Duration.seconds(timeout),
        });

        lambda.addToRolePolicy(
            new PolicyStatement({
                effect: Effect.ALLOW,
                actions: policyActions,
                resources: policiyResources,
            })
        );

        this.lambdaIntegration = new LambdaIntegration(lambda);
        this.lambda = lambda;
    }
}
