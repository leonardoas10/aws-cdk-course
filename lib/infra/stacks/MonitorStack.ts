import { Duration, Stack, StackProps } from 'aws-cdk-lib';
import { Alarm, Metric, Unit } from 'aws-cdk-lib/aws-cloudwatch';
import { SnsAction } from 'aws-cdk-lib/aws-cloudwatch-actions';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Topic } from 'aws-cdk-lib/aws-sns';
import { LambdaSubscription } from 'aws-cdk-lib/aws-sns-subscriptions';
import { Construct } from 'constructs';

interface IMonitorStackProps extends StackProps {
    readonly lambdaFunction: NodejsFunction;
    readonly snsTopicAlarmName: string;
    readonly cloudWatchAlarmName: string;
    readonly evaluationPeriods: number;
    readonly threshold: number;
    readonly metricName: string;
    readonly metricNamespace: string;
    readonly metricPeriod: number;
    readonly metricStatistic: 'Sum';
    readonly metricUnit: Unit;
}

/**
 * The `MonitorStack` class is a custom AWS CDK stack that sets up monitoring
 * and alerting for a given AWS Lambda function using CloudWatch Alarms and SNS.
 *
 * This stack creates a CloudWatch Alarm that monitors a specified metric from
 * a given AWS Lambda function. If the alarm is triggered, it sends a notification
 * through an SNS Topic, which can be handled by the same Lambda function or
 * other subscribers.
 *
 * @param scope The scope in which this stack is defined.
 * @param id The identifier for this stack.
 * @param props The properties for configuring the monitoring stack, including:
 *  - `lambdaFunction`: The Lambda function to monitor.
 *  - `snsTopicAlarmName`: The name of the SNS Topic used for notifications.
 *  - `cloudWatchAlarmName`: The name of the CloudWatch Alarm.
 *  - `evaluationPeriods`: The number of periods over which data is compared
 *    to the specified threshold.
 *  - `threshold`: The value against which the specified metric is compared.
 *  - `metricName`: The name of the CloudWatch metric to monitor.
 *  - `metricNamespace`: The namespace of the CloudWatch metric.
 *  - `metricPeriod`: The period over which the specified statistic is applied.
 *  - `metricStatistic`: The statistic to apply to the metric (e.g., `Sum`).
 *  - `metricUnit`: The unit of the metric being monitored.
 */
export class MonitorStack extends Stack {
    constructor(
        scope: Construct,
        id: string,
        {
            snsTopicAlarmName,
            cloudWatchAlarmName,
            lambdaFunction,
            evaluationPeriods,
            threshold,
            metricName,
            metricNamespace,
            metricPeriod,
            metricStatistic,
            metricUnit,
            ...props
        }: IMonitorStackProps
    ) {
        super(scope, id, props);

        const alarmTopic = new Topic(this, snsTopicAlarmName, {
            displayName: snsTopicAlarmName,
            topicName: snsTopicAlarmName,
        });
        alarmTopic.addSubscription(new LambdaSubscription(lambdaFunction));

        const alarm = new Alarm(this, cloudWatchAlarmName, {
            alarmName: cloudWatchAlarmName,
            evaluationPeriods,
            threshold,
            metric: new Metric({
                metricName,
                namespace: metricNamespace,
                period: Duration.minutes(metricPeriod),
                statistic: metricStatistic,
                unit: metricUnit,
            }),
        });

        const topicAction = new SnsAction(alarmTopic);
        alarm.addAlarmAction(topicAction);
        alarm.addOkAction(topicAction);
    }
}
