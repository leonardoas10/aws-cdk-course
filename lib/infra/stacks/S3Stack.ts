import { Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';

interface IS3BucketStackProps extends StackProps {
    readonly bucketName: string;
    readonly lambda?: NodejsFunction;
}

export class S3BucketStack extends Stack {
    public readonly bucketName: string;
    constructor(
        scope: Construct,
        id: string,
        { bucketName, lambda, ...props }: IS3BucketStackProps
    ) {
        super(scope, id, props);

        const bucket = new Bucket(this, bucketName, {
            bucketName,
        });

        lambda && bucket.grantWrite(lambda);

        bucketName = bucket.bucketName;
    }
}
