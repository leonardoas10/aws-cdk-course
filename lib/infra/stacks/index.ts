export * from './LambdaStack';
export * from './ApiRestStack';
export * from './S3Stack';
export * from './DataStack';
