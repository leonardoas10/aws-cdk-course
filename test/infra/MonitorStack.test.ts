import { App } from 'aws-cdk-lib';
import { Unit } from 'aws-cdk-lib/aws-cloudwatch';
import { Capture, Match, Template } from 'aws-cdk-lib/assertions';

import { LambdaStack, MonitorStack } from '../../lib/infra/stacks';

describe('Test MonitorStack', () => {
    let app: App;
    let lambdaStack: LambdaStack;
    let monitorStack: MonitorStack;

    beforeEach(() => {
        app = new App({
            outdir: 'cdk.out',
        });
        lambdaStack = new LambdaStack(app, 'MonitorLambdaStack', {
            entryPoint: 'services/lambdas/monitor/handler.ts',
            lambdaName: 'monitor',
            policiyResources: undefined,
            policyActions: undefined,
        });

        monitorStack = new MonitorStack(app, 'MonitorStack', {
            lambdaFunction: lambdaStack.lambda,
            cloudWatchAlarmName: 'monitor-slack-cloudwatch',
            snsTopicAlarmName: 'monitor-slack-topic',
            evaluationPeriods: 1,
            metricName: 'Invocations',
            metricNamespace: 'AWS/Lambda',
            metricPeriod: 60,
            metricStatistic: 'Sum',
            metricUnit: Unit.COUNT,
            threshold: 1,
            dimesionsMapApiName: 'SpaceFinderApi',
        });
    });

    test('Lambda propertiees', () => {
        const lambdaStackTemplate = Template.fromStack(lambdaStack);
        lambdaStackTemplate.hasResourceProperties('AWS::Lambda::Function', {
            Handler: 'index.handler',
            Runtime: 'nodejs20.x',
            Timeout: 30,
        });
    });

    test('SNS properties', () => {
        const monitorStackTemplate = Template.fromStack(monitorStack);
        monitorStackTemplate.hasResourceProperties('AWS::SNS::Topic', {
            DisplayName: 'monitor-slack-topic',
            TopicName: 'monitor-slack-topic',
        });
    });

    test('SNS subscription properties - with almost exact values', () => {
        const monitorStackTemplate = Template.fromStack(monitorStack);

        // const snsTopic = monitorStackTemplate.findResources('AWS::SNS::Topic');
        // const snsTopicName = Object.keys(snsTopic)[0];
        // console.log(snsTopic);

        const lambdaStackTemplate = Template.fromStack(lambdaStack);
        lambdaStackTemplate.hasResourceProperties('AWS::SNS::Subscription', {
            Protocol: 'lambda',
            TopicArn: Match.objectLike({
                'Fn::ImportValue': Match.stringLikeRegexp('MonitorStack'),
            }),
            Endpoint: Match.objectLike({
                'Fn::GetAtt': [
                    Match.stringLikeRegexp('monitor'), // matches the logical ID prefix
                    'Arn',
                ],
            }),
        });
    });

    test('Alarm actions - using Capture', () => {
        const monitorStackTemplate = Template.fromStack(monitorStack);
        const alarmActionsCapture = new Capture();
        monitorStackTemplate.hasResourceProperties('AWS::CloudWatch::Alarm', {
            AlarmActions: alarmActionsCapture,
        });

        expect(alarmActionsCapture.asArray()).toEqual([
            {
                Ref: expect.stringMatching(/^monitorslacktopic/),
            },
        ]);
    });

    test('Monitor stack snapshot', () => {
        const monitorStackTemplate = Template.fromStack(monitorStack);
        expect(monitorStackTemplate.toJSON()).toMatchSnapshot();
    });

    test('Lambda stack snapshot', () => {
        const lambdaStackTemplate = Template.fromStack(lambdaStack);
        expect(lambdaStackTemplate.toJSON()).toMatchSnapshot();
    });
});
