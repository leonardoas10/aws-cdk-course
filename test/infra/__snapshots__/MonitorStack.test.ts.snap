// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test MonitorStack Monitor stack snapshot 1`] = `
{
  "Outputs": {
    "ExportsOutputRefmonitorslacktopic40CF14C73449DC5B": {
      "Export": {
        "Name": "MonitorStack:ExportsOutputRefmonitorslacktopic40CF14C73449DC5B",
      },
      "Value": {
        "Ref": "monitorslacktopic40CF14C7",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "monitorslackcloudwatch03AF8A48": {
      "Properties": {
        "AlarmActions": [
          {
            "Ref": "monitorslacktopic40CF14C7",
          },
        ],
        "AlarmName": "monitor-slack-cloudwatch",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "ApiName",
            "Value": "SpaceFinderApi",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "Invocations",
        "Namespace": "AWS/Lambda",
        "OKActions": [
          {
            "Ref": "monitorslacktopic40CF14C7",
          },
        ],
        "Period": 3600,
        "Statistic": "Sum",
        "Threshold": 1,
        "Unit": "Count",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "monitorslacktopic40CF14C7": {
      "Properties": {
        "DisplayName": "monitor-slack-topic",
        "TopicName": "monitor-slack-topic",
      },
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
